src/core/qgsarrayutils.cpp

```diff
@@ -0,0 +1,94 @@
+/***************************************************************************
+                                  qgsarrayutils.h
+                              ---------------------
+    begin                : July 2019
+    copyright            : (C) 2019 by David Signer
+    email                : david at opengis dot ch
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsarrayutils.h"
+#include <QDebug>
+#include <nlohmann/json.hpp>
+using json = nlohmann::json;
+
+QVariantList QgsArrayUtils::parse( const QString &string )
+{
+  QVariantList variantList;
+
+  QString newVal = string;
+  if ( newVal.trimmed().startsWith( '{' ) )
+  {
+    newVal =  newVal.trimmed().mid( 1 ).mid( 0, newVal.length() - 2 ).prepend( '[' ).append( ']' );
+
+    if ( !json::accept( newVal.toStdString() ) )
```

Comment:
Is this what postgres does? I.e. is the postgres quoting arrays the same style as JSON? And nested arrays and maps are expected in JSON structure?

Reply:
I adapted this way of work from the original `QgsValueRelationWidgetFormatter` function to decode it as a first step. I did not replace the `QgsPostgresConn::quotedList` and `QgsPostgresProvider::parseArray``yet.

Reply:
I see. This work was introduced at the same time as the de(serialization) code in this widget was removed, so it's quite young too.

Authoritative here should be how postgres handles this (and not how our postgres provider implements it or how QGIS 3.8 thinks it should be done). Let me have a look ;)

Reply:
Some random tests on postgres

| string                          | result                       |
|---------------------------------|------------------------------|
| SELECT '{1, 2}'::int[]          | [1, 2]                       |
| SELECT '{a, b}'::text[]         | ['a', 'b']                   |
| SELECT '{"a", "b"}'::text[]     | ['a', 'b']                   |
| SELECT '{ "a\"", "b" }'::text[] | ['a"', 'b']                  |
| SELECT '{a [], b}'::text[]      | ['a []', 'b']                |
| SELECT '{a {}, b}'::text[]      | UB / Malformed array literal |
| SELECT '{a \, b}'::text[]       | ['a , b']                    |

Reply:
In the end it's: write strings in quotes and escape doublequotes in a strings.